### PR TITLE
Add a new test test_csi_expansion_canceling()

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#978

### CODE REVIEW

File: manager/integration/tests/test_basic.py [hunk 1]
```diff
@@ -1794,3 +1796,92 @@ def test_engine_image_daemonset_restart(clients, apps_api, volume_name):  # NOQA
...
+    4. Delete the pod and wait for volume detachment
+    5. Try to expand the volume using Longhorn API
+    6. Wait for expansion failure then use Longhorn API to cancel it
+    7. Create a new pod and validate the volume content,
+       then re-write random data to the pod
+    8. Delete the pod and wait for volume detachment
+    9. Retry expansion then verify the expansion done using Longhorn API
+    10. Create a new pod
+    11. Validate the volume content, then check if data writing looks fine
+    12. Cleam up pod, PVC, and PV
```
Comment (yasker) [inline manager/integration/tests/test_basic.py:1819]: Typo.

File: manager/integration/tests/test_basic.py [hunk 2]
```diff
@@ -1794,3 +1796,92 @@ def test_engine_image_daemonset_restart(clients, apps_api, volume_name):  # NOQA
     snap2_data = write_volume_random_data(volume)
     create_snapshot(client, volume_name)
     check_volume_data(volume, snap2_data)
+
+
+@pytest.mark.coretest  # NOQA
+def test_expansion_canceling(clients, core_api, volume_name, pod):  # NOQA
+    """
+    Test expansion canceling feature
```
Comment (yasker) [inline manager/integration/tests/test_basic.py:1804]: Remove `feature`. It's a part of expansion feature.

File: manager/integration/tests/test_ha.py [hunk 3]
```diff
@@ -708,3 +712,82 @@ def test_rebuild_failure_with_intensive_data(client, core_api, volume_name, pod_
...
+    2. Generate `test_data` and write to the pod
+    3. Create an empty directory with expansion snapshot tmp meta file path
+       for one replica so that the replica expansion will fail
+    4. Delete the pod and wait for volume detachment
+    5. Expand the volume using Longhorn API
+    6. Check expansion result using Longhorn API. There will be expansion error
+       caused by the failed replica but overall the expansion should succeed.
+    7. Create a new pod and
+       check if the volume will rebuild for the failed replica
+    8 .Validate the volume content, then check if data writing looks fine
```
Comment (yasker) [inline manager/integration/tests/test_ha.py:732]: Wrong position for `.`

**Comment by shuo-wu** [comment]
1. As long as the PVC spec.size is expanded, we cannot shrink it anymore. Hence the expansion canceling cannot be triggered via PVC/CSI.
2. If users choose PVC to expand the volume then try to use Longhorn API to cancel the expansion, it would lead to the inconsistency between PVC and Longhorn volume. 

Hence expansion canceling feature is mainly for the users using the Longhorn API to expand volumes. And we need to add a tip for the API on the UI page.

**Comment by yasker** [comment]
@shuo-wu How about one replica failure when expansion? The expansion should succeed, right? That's something we should test with CSI as well.

**Comment by shuo-wu** [comment]
Yes. That's pretty similar to the 2nd test in this PR. Do we still need to retain the 2nd test after adding the CSI test you mentioned?

**Comment by yasker** [comment]
It's better to do it through the CSI if we only want to keep one case.

**Pull request merged**